content/asm_5.md

```diff
@@ -1,138 +1,449 @@
+        ;; Set the first argument of `sys_exit` to the first argument of the macro.
+        mov rdi, %1
+        ;; Call the `sys_exit` system call.
         syscall
 %endmacro
 ```
 
-Let's try to go through it macro and understand how it works: At first line we defined PRINT macro with one parameter. Than we push all general registers (with pusha instruction) and flag register with (with pushf instruction). After this we jump to %%astr label. Pay attention that all labels which defined in macro must start with %%. Now we move to __syscall_write macro with 2 parameter. Let's look on __syscall_write implementation. You can remember that we use write system call in all previous posts for printing string to stdout. It looks like this:
+After we have defined our macros, we can use them in our code. For example:
 
 ```assembly
-;; write syscall number
-mov rax, 1
-;; file descriptor, standard output
-mov rdi, 1
-;; message address
-mov rsi, msg
-;; length of message
-mov rdx, 14
-;; call write syscall
-syscall
+;; Definition of the .data section.
+section .data
+        ;; The first message to print.
+        msg_1   db      "Message 1"
+        ;; The second message to print.
+        msg_2   db      "Message 2"
+        ;; ASCII code of the new line symbol ('\n').
+        newline db      0xA
+
+;; Definition of the .text section.
+section .text
+        ;; Reference to the entry point of our program.
+        global _start
+
+;; Entry point of the program.
+_start:
+        ;; Print the first message with the length 9.
+        PRINT msg_1, 9
+        ;; Print new line message with the length 1.
+        PRINT newline, 1
+        ;; Print the second message with the length 9.
+        PRINT msg_2, 9
+        ;; Print new line message with the length 1.
+        PRINT newline, 1
+        ;; Exit from the program. The 0 status code is success.
+        EXIT 0
+```
+
+Now let's go through the macro definitions and see how they work. A macro definition starts from the `%macro` directive, followed by the macro’s name and the number of input parameters it expects. For example, our `PRINT` macro expects two arguments:
+
+- The reference to the string that is going to be printed
+- The length of this string
+
+The `EXIT` macro expects only one input argument - exit code. 
+
+There is also an additional syntax (`n-*`) that indicates a macro accepts at least `n` arguments.
+
+In the macro's body, we initialize the registers according to the [ABI](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf) to make [system calls](https://en.wikipedia.org/wiki/System_call). The only difference you may notice is that instead of using direct values, we use input parameters. These parameters start with the `%` symbol, followed by the number of the parameter (`%1`, `%2`, etc).
+
+Another syntax ability in the multi-line NASM macros is the possibility to define labels inside a macro. In this case, the label name should be prefixed with `%%`. For example:
```

Comment:
Should we mention why anyone would need to define labels? Or is it rather obvious?